storage/innobase/trx/trx0rseg.cc

```diff
@@ -584,7 +584,17 @@ dberr_t trx_rseg_array_init()
 #endif
 	mtr_t mtr;
 	dberr_t err = DB_SUCCESS;
-
+	/* mariabackup --prepare only deals with the redo log and the data
+	files, not with	transactions or the data dictionary, that's why
+	trx_lists_init_at_db_start() does not invoke purge_sys.create() and
+	purge queue mutex stays uninitialized, and trx_rseg_mem_restore() quits
+	before initializing undo log lists. */
+	if (srv_operation != SRV_OPERATION_RESTORE)
+		/* Acquiring purge queue mutex here should be fine from the
+		deadlock prevention point of view, because executing that
+		function is a prerequisite for starting the purge subsystem or
+		any transactions. */
+		purge_sys.queue_lock();
```

Comment:
Omitting the condition would result in a smaller executable. If there is no contention on the mutex (like there should not be here), unconditionally acquiring and releasing the mutex should be no problem.

Reply:
But the mutex is not initialized for `srv_operation == SRV_OPERATION_RESTORE`.